### PR TITLE
More rdoc for ENV

### PR DESCRIPTION
**Pull request by BurdetteLamar**

### CODE REVIEW

File: hash.c [hunk 1]
```diff
@@ -6214,6 +6214,33 @@ Init_Hash(void)
...
+     * - When you create or set a name-value pair in ENV, the name and value are immediately set in the environment variables.
+     * - When you delete a name-value pair in ENV, it is immediately deleted from the environment variables.
+     * === Names and Values
+     * Generally speaking, each name or value is a String.
+     *
+     * Strictly speaking:
+     * - Each name or value must be one of the following:
+     *   - A String.
+     *   - An object that responds to #to_str by returning a String, which will be used as the name or value.
+     * - A name may not:
```
Comment (nobu) [inline hash.c]: Also a name and value may not contain NUL character.

File: hash.c [hunk 2]
```diff
@@ -6214,6 +6214,33 @@ Init_Hash(void)
...
+     *   - Contain character <code>=</code>.
+     *   - Be the empty string.
+
+     * === About Ordering
+     * The ordering of ENV content is OS-dependent.
```
Comment (nobu) [inline hash.c]: Not only OS-dependent, it is indeterminate.

Comment (BurdetteLamar) [reply hash.c]: I  have not seen indeterminate in Windows or Ubuntu.  But will add.  Should I say "OS-dependent and possibly indeterminate"?

File: hash.c (hunk 1 above)
Comment (BurdetteLamar) [reply hash.c]: I saw nothing about that in spec.  Will add to spec soon.  And will add to hash.c doc.

Comment (BurdetteLamar) [reply hash.c]: Added here.

File: hash.c (hunk 2 above)
Comment (BurdetteLamar) [reply hash.c]: Added here.

File: hash.c (hunk 1 above)
Comment (nobu) [reply hash.c]: `INVALID_ENVVARS` in test/ruby/test_env.rb are examples.

File: hash.c [hunk 3]
```diff
@@ -6214,6 +6214,33 @@ Init_Hash(void)
...
+     * - When you delete a name-value pair in ENV, it is immediately deleted from the environment variables.
+     * === Names and Values
+     * Generally speaking, each name or value is a String.
+     *
+     * Strictly speaking:
+     * - Each name or value must be one of the following:
+     *   - A String.
+     *   - An object that responds to #to_str by returning a String, which will be used as the name or value.
+     * - A name may not:
+     *   - Contain character <code>=</code> or the NUL character (<code>"\0"</code>
```
Comment (nobu) [inline hash.c]: Missing ')'?

File: hash.c (hunk 1 above)
Comment (BurdetteLamar) [reply hash.c]: Do we need to say more about the encodings?

**Pull request merged**